service/history/replication/stream.go

```diff
@@ -56,3 +59,21 @@ func ClusterIDToClusterNameShardCount(
 	}
 	return "", 0, serviceerror.NewInternal(fmt.Sprintf("unknown cluster ID: %v", clusterID))
 }
+
+func GetErrorHandledAndRetriedEventLoop(originalEventLoop func() error, streamStopper func(), logger log.Logger, metricsHandler metrics.Handler) func() {
+	return func() {
+		defer streamStopper()
+		for {
+			err := originalEventLoop()
+			if err != nil {
+				if _, ok := err.(*StreamError); ok {
+					metricsHandler.Counter(metrics.ReplicationStreamError.Name()).Record(1)
+					return
+				}
+				time.Sleep(streamReceiverMonitorInterval)
```

Comment:
this is new behavior right? can you entail?

Reply:
This is actually not new behavior. Currently, when there is any error returned from event loop, the stream will be terminated. Then the stream will be re-established by [stream monitor](https://github.com/temporalio/temporal/blob/6a5a86b9000b2fa4e2a62ca3c36258e03674ebec/service/history/replication/stream_receiver_monitor.go#L136). The interval of this monitor is 2 seconds. So The sleep here is effectively the same.

Reply:
so this is an optimization to reduce # of stream re-establish? took a quick look at the stream monitor code, seems new stream will be established every 2s?

Reply:
Yes this is an optimization. In original code, if there is any error no matter stream error or business logic error, the eventloop will exit and mark stream as **invalid**. Then stream_receiver_monitor will re-establish it in 2 seconds. My code here is to distinguish stream error and make sure only markthe stream **invalid** when we see stream error. 

`seems new stream will be established every 2s` it check every 2 seconds, if stream status are valid, no new stream will be established.

Reply:
thanks for the explanation. would nice to document them in the code.